### PR TITLE
Use IsEditorRequired in Equality checks

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 namespace Microsoft.AspNetCore.Razor.Language
 {
-    internal class BoundAttributeDescriptorComparer : IEqualityComparer<BoundAttributeDescriptor>
+    internal sealed class BoundAttributeDescriptorComparer : IEqualityComparer<BoundAttributeDescriptor>
```
Comment (SteveSandersonMS) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:11]: Seeing how much `sealed` shows up in @pranavkm's PRs, I propose renaming it to "the Pranav modifier".

**Review by SteveSandersonMS — APPROVED** [review]
I guess you're fixing something specific here, but I can't guess what. I'm interested to know, but either way, the changes look totally reasonable.

**Comment by pranavkm** [comment]
> https://github.com/dotnet/aspnetcore/pull/33553

I discovered it as part of https://github.com/dotnet/aspnetcore-tooling/pull/3798/files. I had initially thought that the property would necessarily matter for comparison, but there's some sort of caching that is performed by tooling: https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperDescriptorCache.cs#L8 which tripped up a test.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs (hunk 1 above)
Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:11]: #pranavPoints 😄

**Review by TanayParikh — APPROVED** [review]

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs [hunk 2]
```diff
@@ -48,7 +49,7 @@ public virtual bool Equals(BoundAttributeDescriptor descriptorX, BoundAttributeD
                     descriptorY.Metadata.OrderBy(propertyY => propertyY.Key, StringComparer.Ordinal));
         }
 
-        public virtual int GetHashCode(BoundAttributeDescriptor descriptor)
+        public int GetHashCode(BoundAttributeDescriptor descriptor)
         {
             if (descriptor == null)
```
Comment (TanayParikh) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: Potentially worth adding this to ensure the TagHelper caching mechanism is able to differentiate between two similar taghelpers with different `IsEditorRequired` values.

```
hash.Add(descriptor.IsEditorRequired);
```

Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: Yep, you'll need this to get around the [caching](https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/TagHelperDescriptorCache.cs#L8) as it's based on hash code and not equality. Disabling auto-merge.

Comment (pranavkm) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: Oh fun, I missed that. Do you know how it works for properties like `IsEnum` that aren't enumerated here?

Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: I'd recommend adding in all uniquely identifiable properties into this GetHashCode. I tried to strike a balance between identifiable properties and simplicity (to ensure hashing doesn't add any substantial cost).

Hashing PR: https://github.com/dotnet/aspnetcore/pull/24551
Perf Benchmarks: https://github.com/dotnet/aspnetcore-tooling/pull/2307

Comment (pranavkm) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: But isn't the caching already busted at that point? In general though, it seems very problematic to use a hash to cache items.  A hash isn't guaranteed to be unique. If tooling wanted to ensure this behavior, it should write it's own comparer that guarantees it's requirements.

Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: Hashing is imperfect, but it was the tradeoff between accuracy and perf. Assuming enough entropy in the hash (by covering the subset of unique properties) chances of a hash collision should be minimal. Without this caching each `GetTagHelpers` call through Roslyn is going to cost ~230 tag helper allocations. And this is a fairly frequent call (IIRC occurs during compilations).

Going the comparator route would ensure accuracy but would undermine much of the perf benefits (as we'd still need to do a bunch of allocations and so on in order to subsequently do the comparisons on the tooling end).

Comment (pranavkm) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/BoundAttributeDescriptorComparer.cs:54]: Right - but you get incorrect results with what have you now. Correctness > perf.

**Pull request merged**